File: src/Daisi.Llogos.WebGpu/src/tokenizer/bpe-tokenizer.ts
/**
 * BPE tokenizer — ported from Daisi.Llogos.Tokenizer.BpeTokenizer.cs
 * Supports both GPT-2 byte encoding and direct UTF-8 (Qwen/Llama 3).
 */

// GPT-2 pre-tokenization regex
const PRE_TOKENIZE_RE = /'(?:[sdmt]|ll|ve|re)|[^\r\n\p{L}\p{N}]?\p{L}+|\p{N}{1,3}| ?[^\s\p{L}\p{N}]+[\r\n]*|\s*[\r\n]+|\s+(?!\S)|\s+/gu;

export class BpeTokenizer {
  private tokens: string[];
  private tokenToId: Map<string, number>;
  private mergeRank: Map<string, number>;
  private useByteEncoding: boolean;
  private specialTokens: string[]; // sorted longest-first for greedy matching
  readonly bosTokenId: number;
  readonly eosTokenId: number;
  readonly padTokenId: number;

  constructor(
    tokens: string[],
    merges: string[],
    bosTokenId: number,
    eosTokenId: number,
    padTokenId: number,
    useByteEncoding: boolean,
  ) {
    this.tokens = tokens;
    this.bosTokenId = bosTokenId;
    this.eosTokenId = eosTokenId;
    this.padTokenId = padTokenId;
    this.useByteEncoding = useByteEncoding;

    this.tokenToId = new Map();
    for (let i = 0; i < tokens.length; i++) {
      if (!this.tokenToId.has(tokens[i])) {
        this.tokenToId.set(tokens[i], i);
      }
    }

    this.mergeRank = new Map();
    for (let i = 0; i < merges.length; i++) {
      this.mergeRank.set(merges[i], i);
    }

    // Collect special tokens (tokens with < > or similar patterns)
    this.specialTokens = tokens
      .filter(t => t.startsWith('<') && t.endsWith('>') && t.length > 2)
      .sort((a, b) => b.length - a.length); // longest first for greedy match
  }

  get vocabSize(): number { return this.tokens.length; }

  /** Get token ID for a string, or -1 if not found. */
  getTokenId(token: string): number {
    return this.tokenToId.get(token) ?? -1;
  }

  /** Encode text to token IDs, handling special tokens. */
  encode(text: string): number[] {
    if (!text) return [];

    // Split text around special tokens, then BPE-encode the non-special parts
    const segments = this.splitOnSpecialTokens(text);
    const result: number[] = [];

    for (const seg of segments) {
      if (seg.isSpecial) {
        const id = this.tokenToId.get(seg.text);
        if (id !== undefined) result.push(id);
        // Also try with fullwidth pipes (Qwen style)
        else {
          const fwToken = seg.text.replace(/\|/g, '\uFF5C');
          const fwId = this.tokenToId.get(fwToken);
          if (fwId !== undefined) result.push(fwId);
        }
      } else {
        result.push(...this.encodeBpe(seg.text));
      }
    }

    return result;
  }

  /** BPE-encode a text segment (no special tokens). */
  private encodeBpe(text: string): number[] {
    if (!text) return [];
    const result: number[] = [];
    const matches = text.matchAll(PRE_TOKENIZE_RE);

    for (const match of matches) {
      const chunk = match[0];
      const symbols = this.useByteEncoding
        ? byteEncodeChunk(chunk)
        : this.directEncodeChunk(chunk);
      if (symbols.length === 0) continue;

      this.applyMerges(symbols);

      for (const symbol of symbols) {
        const id = this.tokenToId.get(symbol);
        if (id !== undefined) result.push(id);
      }
    }

    return result;
  }

  /** Split text into segments of special tokens and regular text. */
  private splitOnSpecialTokens(text: string): Array<{ text: string; isSpecial: boolean }> {
    const segments: Array<{ text: string; isSpecial: boolean }> = [];
    let remaining = text;

    while (remaining.length > 0) {
      // Find the earliest special token match
      let bestIdx = remaining.length;
      let bestToken = '';

      for (const st of this.specialTokens) {
        const idx = remaining.indexOf(st);
        if (idx >= 0 && idx < bestIdx) {
          bestIdx = idx;
          bestToken = st;
        }
      }

      if (bestToken) {
        // Add text before the special token
        if (bestIdx > 0) {
          segments.push({ text: remaining.slice(0, bestIdx), isSpecial: false });
        }
        // Add the special token
        segments.push({ text: bestToken, isSpecial: true });
        remaining = remaining.slice(bestIdx + bestToken.length);
      } else {
        // No more special tokens — add remaining text
        segments.push({ text: remaining, isSpecial: false });
        break;
      }
    }

    return segments;
  }

  /** Decode token IDs back to text. */
  decode(tokenIds: number[]): string {
    const parts: string[] = [];
    for (const id of tokenIds) {
      if (id === this.bosTokenId || id === this.eosTokenId || id === this.padTokenId) continue;

      let token = this.tokens[id];
      if (!token) continue;

      // Handle byte fallback tokens like <0x0A>
      if (token.length === 6 && token.startsWith('<0x') && token.endsWith('>')) {
        const byte = parseInt(token.slice(3, 5), 16);
        if (!isNaN(byte)) {
          parts.push(String.fromCharCode(byte));
          continue;
        }
      }

      // Normalize fullwidth pipes (Qwen uses ｜ in special tokens)
      if (token.includes('\uFF5C')) {
        token = token.replaceAll('\uFF5C', '|');
      }

      parts.push(token);
    }

    let text = parts.join('');
    if (this.useByteEncoding) {
      return byteDecodeString(text);
    }
    // SentencePiece space marker
    return text.replaceAll('\u2581', ' ');
  }

  /** Check if a token ID is an end-of-sequence token. */
  isEos(tokenId: number): boolean {
    if (tokenId === this.eosTokenId) return true;
    // Some models have multiple EOS tokens
    const token = this.tokens[tokenId];
    return token === '<|endoftext|>' || token === '<|im_end|>'
      || token === '<｜end▁of▁text｜>' || token === '<｜im_end｜>';
  }

  // Direct encoding for SentencePiece/Llama style vocabs
  private directEncodeChunk(chunk: string): string[] {
    const symbols: string[] = [];
    const encoder = new TextEncoder();
    const bytes = encoder.encode(chunk);
    let i = 0;
    while (i < bytes.length) {
      const charLen = utf8CharLength(bytes[i]);
      if (i + charLen <= bytes.length) {
        const ch = new TextDecoder().decode(bytes.slice(i, i + charLen));
        if (this.tokenToId.has(ch)) {
          symbols.push(ch);
          i += charLen;
          continue;
        }
      }
      symbols.push(`<0x${bytes[i].toString(16).toUpperCase().padStart(2, '0')}>`);
      i++;
    }
    return symbols;
  }

  // BPE merge algorithm
  private applyMerges(symbols: string[]): void {
    while (symbols.length > 1) {
      let bestRank = Infinity;
      let bestIdx = -1;

      for (let i = 0; i < symbols.length - 1; i++) {
        const key = `${symbols[i]} ${symbols[i + 1]}`;
        const rank = this.mergeRank.get(key);
        if (rank !== undefined && rank < bestRank) {
          bestRank = rank;
          bestIdx = i;
        }
      }

      if (bestIdx < 0) break;
      symbols[bestIdx] = symbols[bestIdx] + symbols[bestIdx + 1];
      symbols.splice(bestIdx + 1, 1);
    }
  }
}

// ── GPT-2 byte encoding ─────────────────────────────────────────────

const BYTE_TO_UNICODE = buildByteToUnicode();
const UNICODE_TO_BYTE = buildUnicodeToByte();

function buildByteToUnicode(): string[] {
  const table = new Array<string>(256);
  let n = 256;
  for (let i = 0; i < 256; i++) {
    if ((i >= 0x21 && i <= 0x7E) || (i >= 0xA1 && i <= 0xAC) || (i >= 0xAE && i <= 0xFF)) {
      table[i] = String.fromCharCode(i);
    } else {
      table[i] = String.fromCharCode(n);
      n++;
    }
  }
  return table;
}

function buildUnicodeToByte(): Map<string, number> {
  const map = new Map<string, number>();
  for (let i = 0; i < 256; i++) {
    map.set(BYTE_TO_UNICODE[i], i);
  }
  return map;
}

function byteEncodeChunk(chunk: string): string[] {
  const encoder = new TextEncoder();
  const bytes = encoder.encode(chunk);
  return Array.from(bytes, b => BYTE_TO_UNICODE[b]);
}

function byteDecodeString(text: string): string {
  const bytes: number[] = [];
  for (const ch of text) {
    const b = UNICODE_TO_BYTE.get(ch);
    if (b !== undefined) {
      bytes.push(b);
    } else {
      const encoder = new TextEncoder();
      bytes.push(...encoder.encode(ch));
    }
  }
  return new TextDecoder().decode(new Uint8Array(bytes));
}

function utf8CharLength(firstByte: number): number {
  if (firstByte < 0x80) return 1;
  if (firstByte < 0xC0) return 1;
  if (firstByte < 0xE0) return 2;
  if (firstByte < 0xF0) return 3;
  return 4;
}

/**
 * Build a BpeTokenizer from GGUF metadata.
 */
export function tokenizerFromGguf(metadata: Map<string, unknown>): BpeTokenizer {
  const tokens = metadata.get('tokenizer.ggml.tokens') as string[];
  const merges = metadata.get('tokenizer.ggml.merges') as string[];
  if (!tokens || !merges) throw new Error('Missing tokenizer metadata in GGUF');

  const bosTokenId = (metadata.get('tokenizer.ggml.bos_token_id') as number) ?? -1;
  const eosTokenId = (metadata.get('tokenizer.ggml.eos_token_id') as number) ?? -1;
  const padTokenId = (metadata.get('tokenizer.ggml.padding_token_id') as number) ?? -1;
  const model = metadata.get('tokenizer.ggml.model') as string | undefined;
  const useByteEncoding = model === 'gpt2';

  return new BpeTokenizer(tokens, merges, bosTokenId, eosTokenId, padTokenId, useByteEncoding);
}

File: src/dotnet/Daisi.Llogos/Tokenizer/BpeTokenizer.cs
using System.Text;
using System.Text.RegularExpressions;

namespace Daisi.Llogos.Tokenizer;

/// <summary>
/// Byte Pair Encoding tokenizer. Encodes text to token IDs and decodes back.
/// Uses a vocabulary and merge table extracted from GGUF metadata.
/// Supports both GPT-2 byte encoding and direct UTF-8 token vocabularies (e.g. Qwen).
/// </summary>
public sealed partial class BpeTokenizer
{
    private readonly Vocabulary _vocab;
    private readonly MergeTable _merges;
    private readonly bool _useByteEncoding;
    private readonly bool _useSentencePiece;
    private readonly Regex _preTokenizePattern;

    /// <summary>SentencePiece "lower one eighth block" — used as the in-vocab space marker.</summary>
    private const char SpmSpaceMarker = '\u2581';

    // GPT-2 byte-to-token mapping (only used when _useByteEncoding is true)
    private static readonly string[] ByteToUnicodeTable = BuildByteToUnicode();
    private static readonly Dictionary<char, byte> UnicodeToByte = BuildUnicodeToByte();

    // Special tokens that should be matched as whole strings (e.g. <|im_start|>, <|endoftext|>)
    private readonly List<(string token, int id)> _specialTokens = new();

    public BpeTokenizer(Vocabulary vocab, MergeTable merges, bool useByteEncoding = false, bool useSentencePiece = false)
    {
        _vocab = vocab;
        _merges = merges;
        _useByteEncoding = useByteEncoding;
        _useSentencePiece = useSentencePiece;
        _preTokenizePattern = PreTokenizeRegex();

        // Build special token list from vocabulary (tokens matching <|...|> or <...> patterns)
        for (int i = 0; i < vocab.Count; i++)
        {
            var token = vocab.IdToToken(i);
            if (token != null && token.StartsWith("<|") && token.EndsWith("|>") && token.Length > 4)
                _specialTokens.Add((token, i));
            // Also handle fullwidth pipe variants (Qwen uses ｜ in GGUF)
            else if (token != null && token.Contains('\uFF5C') && token.StartsWith("<"))
            {
                var normalized = token.Replace('\uFF5C', '|');
                _specialTokens.Add((normalized, i));
            }
        }
        // Sort by length descending so longer tokens match first
        _specialTokens.Sort((a, b) => b.token.Length.CompareTo(a.token.Length));
    }

    /// <summary>The vocabulary used by this tokenizer.</summary>
    public Vocabulary Vocabulary => _vocab;

    /// <summary>
    /// Encode text into a sequence of token IDs.
    /// </summary>
    public int[] Encode(string text)
    {
        if (string.IsNullOrEmpty(text))
            return [];

        var result = new List<int>();

        // Split on special tokens first, then BPE-encode the segments between them
        var segments = SplitOnSpecialTokens(text);
        foreach (var (segment, specialId) in segments)
        {
            if (specialId >= 0)
            {
                result.Add(specialId);
                continue;
            }

            if (_useSentencePiece)
            {
                // SentencePiece (Gemma, Llama): replace ASCII spaces with U+2581 ('▁'),
                // skip the GPT-2 regex pre-tokenization (which would split on spaces and
                // prevent in-word merges), and BPE-encode the entire segment as one chunk.
                var spmText = SentencePiecePreprocess(segment);
                if (spmText.Length == 0) continue;
                var symbols = DirectEncodeChunk(spmText);
                if (symbols.Count == 0) continue;
                ApplyMerges(symbols);
                foreach (var symbol in symbols)
                {
                    int id = _vocab.TokenToId(symbol);
                    if (id >= 0) result.Add(id);
                }
                continue;
            }

            // Regular BPE encoding for non-special text
            var matches = _preTokenizePattern.Matches(segment);
            foreach (Match match in matches)
            {
                var chunk = match.Value;
                var symbols = _useByteEncoding ? ByteEncodeChunk(chunk) : DirectEncodeChunk(chunk);
                if (symbols.Count == 0) continue;

                ApplyMerges(symbols);

                foreach (var symbol in symbols)
                {
                    int id = _vocab.TokenToId(symbol);
                    if (id >= 0)
                        result.Add(id);
                }
            }
        }

        return result.ToArray();
    }

    /// <summary>
    /// SentencePiece preprocessing: replace each ASCII space with U+2581 ('▁').
    /// This is the convention SPM-trained models (Gemma, Llama) use to represent
    /// word boundaries inside the vocabulary.
    /// </summary>
    private static string SentencePiecePreprocess(string text)
    {
        if (text.Length == 0) return text;
        var sb = new StringBuilder(text.Length);
        foreach (var ch in text)
            sb.Append(ch == ' ' ? SpmSpaceMarker : ch);
        return sb.ToString();
    }

    /// <summary>
    /// Split text into segments, identifying special tokens as whole units.
    /// Returns (text_segment, -1) for regular text, (token_text, token_id) for special tokens.
    /// </summary>
    private List<(string text, int id)> SplitOnSpecialTokens(string text)
    {
        if (_specialTokens.Count == 0)
            return [(text, -1)];

        var segments = new List<(string text, int id)>();
        int pos = 0;

        while (pos < text.Length)
        {
            // Try to match a special token at current position
            bool found = false;
            foreach (var (token, id) in _specialTokens)
            {
                if (pos + token.Length <= text.Length &&
                    text.AsSpan(pos, token.Length).SequenceEqual(token.AsSpan()))
                {
                    // Flush any preceding text
                    if (pos > 0 && segments.Count == 0 || (segments.Count > 0 && segments[^1].id < 0))
                    {
                        // Check if there's text before this match not yet added
                    }
                    // Simpler approach: collect start position
                    segments.Add((token, id));
                    pos += token.Length;
                    found = true;
                    break;
                }
            }
            if (!found)
            {
                // Accumulate regular text
                if (segments.Count > 0 && segments[^1].id < 0)
                {
                    segments[^1] = (segments[^1].text + text[pos], -1);
                }
                else
                {
                    segments.Add((text[pos].ToString(), -1));
                }
                pos++;
            }
        }

        return segments;
    }

    /// <summary>
    /// Decode a sequence of token IDs back to text.
    /// </summary>
    public string Decode(ReadOnlySpan<int> tokenIds)
    {
        var sb = new StringBuilder();
        foreach (var id in tokenIds)
        {
            if (id == _vocab.BosTokenId || id == _vocab.EosTokenId || id == _vocab.PadTokenId)
                continue;

            var token = _vocab.IdToToken(id);

            // Handle byte fallback tokens like <0x0A>
            if (token.Length == 6 && token.StartsWith("<0x") && token.EndsWith('>'))
            {
                if (byte.TryParse(token.AsSpan(3, 2), System.Globalization.NumberStyles.HexNumber, null, out var b))
                {
                    sb.Append((char)b);
                    continue;
                }
            }

            // Normalize fullwidth pipes (U+FF5C ｜) to ASCII pipes (U+007C |)
            // in special tokens. Qwen models store tokens like <｜im_end｜> with
            // fullwidth pipes, but stop sequences and prompts use ASCII pipes.
            if (token.Contains('\uFF5C'))
                token = token.Replace('\uFF5C', '|');

            sb.Append(token);
        }

        var text = sb.ToString();
        if (_useByteEncoding)
            return ByteDecodeString(text);
        // SentencePiece uses ▁ (U+2581) as space marker
        return text.Replace('\u2581', ' ');
    }

    // ── Direct encoding (Qwen, LLaMA 3, etc.) ──────────────────────────────

    /// <summary>
    /// For vocabularies that store tokens as direct UTF-8 strings.
    /// Each character becomes a symbol; unknown chars fall back to byte tokens.
    /// </summary>
    private List<string> DirectEncodeChunk(string chunk)
    {
        var symbols = new List<string>();
        var bytes = Encoding.UTF8.GetBytes(chunk);
        int i = 0;
        while (i < bytes.Length)
        {
            // Try to match the longest single character at this position
            int charLen = GetUtf8CharLength(bytes[i]);
            if (i + charLen <= bytes.Length)
            {
                var ch = Encoding.UTF8.GetString(bytes, i, charLen);
                if (_vocab.Contains(ch))
                {
                    symbols.Add(ch);
                    i += charLen;
                    continue;
                }
            }

            // Fall back to byte token <0xHH>
            symbols.Add($"<0x{bytes[i]:X2}>");
            i++;
        }
        return symbols;
    }

    private static int GetUtf8CharLength(byte firstByte)
    {
        if (firstByte < 0x80) return 1;
        if (firstByte < 0xC0) return 1; // continuation byte (shouldn't be first)
        if (firstByte < 0xE0) return 2;
        if (firstByte < 0xF0) return 3;
        return 4;
    }

    // ── GPT-2 byte encoding ─────────────────────────────────────────────────

    /// <summary>
    /// For GPT-2 style vocabularies that use unicode remapping for bytes.
    /// </summary>
    private static List<string> ByteEncodeChunk(string chunk)
    {
        var bytes = Encoding.UTF8.GetBytes(chunk);
        var symbols = new List<string>(bytes.Length);
        foreach (var b in bytes)
            symbols.Add(ByteToUnicodeTable[b]);
        return symbols;
    }

    private static string ByteDecodeString(string bpeText)
    {
        var bytes = new List<byte>(bpeText.Length);
        foreach (var ch in bpeText)
        {
            if (UnicodeToByte.TryGetValue(ch, out var b))
                bytes.Add(b);
            else
                bytes.AddRange(Encoding.UTF8.GetBytes(ch.ToString()));
        }
        return Encoding.UTF8.GetString(bytes.ToArray());
    }

    // ── BPE merge algorithm ─────────────────────────────────────────────────

    /// <summary>
    /// Apply BPE merges to a list of symbols in priority order.
    /// Iteratively finds and applies the highest-priority (lowest rank) merge.
    /// </summary>
    private void ApplyMerges(List<string> symbols)
    {
        while (symbols.Count > 1)
        {
            int bestRank = int.MaxValue;
            int bestIdx = -1;

            for (int i = 0; i < symbols.Count - 1; i++)
            {
                int rank = _merges.GetRank(symbols[i], symbols[i + 1]);
                if (rank >= 0 && rank < bestRank)
                {
                    bestRank = rank;
                    bestIdx = i;
                }
            }

            if (bestIdx < 0) break;

            symbols[bestIdx] = symbols[bestIdx] + symbols[bestIdx + 1];
            symbols.RemoveAt(bestIdx + 1);
        }
    }

    // ── GPT-2 byte table construction ───────────────────────────────────────

    private static string[] BuildByteToUnicode()
    {
        var table = new string[256];
        int n = 256;
        for (int i = 0; i < 256; i++)
        {
            if ((i >= 0x21 && i <= 0x7E) || (i >= 0xA1 && i <= 0xAC) || (i >= 0xAE && i <= 0xFF))
                table[i] = ((char)i).ToString();
            else
            {
                table[i] = ((char)n).ToString();
                n++;
            }
        }
        return table;
    }

    private static Dictionary<char, byte> BuildUnicodeToByte()
    {
        var table = BuildByteToUnicode();
        var reverse = new Dictionary<char, byte>(256);
        for (int i = 0; i < 256; i++)
            reverse[table[i][0]] = (byte)i;
        return reverse;
    }

    // GPT-2 / Qwen pre-tokenization regex
    [GeneratedRegex(@"'(?:[sdmt]|ll|ve|re)|[^\r\n\p{L}\p{N}]?\p{L}+|\p{N}{1,3}| ?[^\s\p{L}\p{N}]+[\r\n]*|\s*[\r\n]+|\s+(?!\S)|\s+", RegexOptions.Compiled)]
    private static partial Regex PreTokenizeRegex();
}
